File: Trees/LC_110.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.29

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        def cal_height(root):
            if not root:
                return 0
            l = cal_height(root.left)
            r = cal_height(root.right)
            return 1 + max(l, r)
        
        return abs(cal_height(root.left) - cal_height(root.right)) <= 1 and self.isBalanced(root.left) and self.isBalanced(root.right)

File: Trees/LC_110.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.29

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (root == nullptr){
            return true;
        }
        int l = cal_height(root->left), r = cal_height(root->right);
        return (abs(l-r) <= 1) && isBalanced(root->left) && isBalanced(root->right);
    }

private:
    int cal_height(TreeNode* root){
        if (root == nullptr){
            return 0;
        }
        int l = cal_height(root->left), r = cal_height(root->right);
        return 1 + max(l, r);

    }
};